### ISSUE TITLE
Image Search Performance on CPU

### ISSUE THREAD

**Issue by shersoni610** [jina-ai/jina#4103]
Has anyone tried, the image search example on a MAC machine that is available on Jina webpage? On my machine, it is ran for hours without doing anything. 

Thanks

**Comment by JoanFM** [comment]
Hello @shersoni610 , can u please put some more details on what steps you follow and what is the observed outcome?

Also it would be nice if you can set `JINA_LOG_LEVEL` environment variable  to `DEBUG` before running it

Thanks

**Comment by shersoni610** [comment]
Hello, I just copied the example that is available on the webpage and changed cuda to CPU. I used the same image data that was used in the example:

from jina import DocumentArray, Document
import torchvision

def preproc(d: Document):
    return (d.load_uri_to_image_blob()  # load
             .set_image_blob_normalization()  # normalize color 
             .set_image_blob_channel_axis(-1, 0))  # switch color axis
docs = DocumentArray.from_files('img/*.jpg').apply(preproc)

print( "Step 1 " )
model = torchvision.models.resnet50(pretrained=True)  # load ResNet50
print( "Step 2 " )
docs.embed(model, device='cpu')  # embed via GPU to speedup

print( "Step 3 " )

q = (Document(uri='img/00021.jpg')  # build query image & preprocess
     .load_uri_to_image_blob()
     .set_image_blob_normalization()
     .set_image_blob_channel_axis(-1, 0))
q.embed(model)  # embed
q.match(docs)  # find top-20 nearest neighbours, done!

print( "Step 4 " )

for m in q.matches:
    m.set_image_blob_channel_axis(0, -1).set_image_blob_inv_normalization()
q.matches.plot_image_sprites()

On the MAC book pro, it is taking exorbitant time to get to the Step#4.  Wondering
if I am doing wrong.

Thanks

**Comment by JoanFM** [comment]
Hey @shersoni610 ,

May I ask you, fo you have a folder called `img` with actual `.jpg` images in it?=

**Comment by shersoni610** [comment]
Yes. I get the result as shown on the webpage but after an exorbitant time.

**Comment by shersoni610** [comment]
For the MedicalMNIST dataset, it took almost 2 hours for 10000 images.  The classical image search using OpenCV features takes a couple of mins and the results were reasonable.

**Comment by JoanFM** [comment]
What is the step that is taking time?

**Comment by shersoni610** [comment]
Step#2 i.e. docs.embed() is very slow.

**Comment by JoanFM** [comment]
> Step#2 i.e. docs.embed() is very slow.

Well yes, embedding 10000 images using CPU with ResNet can be rather slow

**Comment by shersoni610** [comment]
Is there an example to show how to store the embedding in a JSON file (or some in-memory database)?

**Comment by JoanFM** [comment]
You can always print the document as a json:

```python
docs.to_json()
```

This may be too big given the size of the documentarray.

**Comment by shersoni610** [comment]
Can we store OpenCV features and search for image similarity using JINA?

**Comment by shersoni610** [comment]
Docs may be large but embedding should be small? Am I missing something?

**Comment by JoanFM** [comment]
> Can we store OpenCV features and search for image similarity using JINA?

You can store any `numpy array ` or `torch tensor` as the embedding of a `Document`.

**Comment by JoanFM** [comment]
> Docs may be large but embedding should be small? Am I missing something?

Embeddings would be high-dimensional arrays. You can try and see

**Comment by JoanFM** [comment]
Hello @shersoni610 ,

Is there any progress?
